src/ide/analysis_progress.rs

```diff
@@ -0,0 +1,180 @@
+use std::collections::HashSet;
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::{Arc, Mutex};
+
+use lsp_types::notification::Notification;
+
+use crate::id_generator::IdGenerator;
+use crate::server::client::Notifier;
+use crate::state::Beacon;
+
+/// A facade for `AnalysisProgressController` that allows to track progress of diagnostics
+/// generation and procmacro requests.
+#[derive(Clone)]
+pub struct AnalysisProgressTracker {
+    controller: AnalysisProgressController,
+}
+
+impl AnalysisProgressTracker {
+    /// Signals that a request to proc macro server was made during the current generation of
+    /// diagnostics.
+    pub fn register_procmacro_request(&self) {
+        self.controller.set_did_submit_procmacro_request(true);
+    }
+
+    /// Sets handlers for tracking beacons sent to threads.
+    /// The beacons are wrapping snapshots, which are signalling when diagnostics finished
+    /// calculating for a given snapshot (used for calculating files diagnostics or removing
+    /// stale ones)
+    pub fn track_analysis<'a>(&self, beacons: impl Iterator<Item = &'a mut Beacon>) {
+        let gen_id = self.controller.next_generation_id();
+
+        self.controller.clear_active_snapshots();
+
+        beacons.enumerate().for_each(|(i, beacon)| {
+            self.controller.insert_active_snapshot(i);
+
+            let controller_ref: AnalysisProgressController = self.controller.clone();
+            beacon.on_signal(move || controller_ref.on_snapshot_deactivate(gen_id, i));
+        });
+
+        self.controller.start_analysis();
+    }
+}
+
+/// Controller used to send notifications to the client about analysis progress.
+/// Uses information provided from other controllers (diagnostics controller, procmacro controller)
+/// to assess if diagnostics are in fact calculated.
+#[derive(Debug, Clone)]
+pub struct AnalysisProgressController {
+    notifier: Notifier,
+    /// ID of the diagnostics "generation" - the scheduled diagnostics jobs set.
+    /// Used to filter out stale threads finishing when new ones (from newer "generation")
+    /// are already in progress and being tracked by the controller.
+    generation_id: Arc<AtomicU64>,
+    /// Sequential IDs of state snapshots from the current generation, used to track their status
+    /// (present meaning it's still being used)
+    active_snapshots: Arc<Mutex<HashSet<usize>>>,
+    id_generator: Arc<IdGenerator>,
```

Comment:
TBH you'd be better off not touching the IdGenerator here and just peeing that AtomicU64 in generation_id and incrementing it directly. you're just doing this in a fancy way now